### DISCUSSION TITLE
SessionProvider Wrapper Disappears with NextJS 13's App Router

### DISCUSSION THREAD

**Discussion by lplawlor** [nextauthjs/next-auth#7612] [Help]
Hi, I'm just learning Next and NextAuth and I'm hoping someone might better understand what's going on here.

I'm building an app using Next 13's App Router and NextAuth for authentication through Spotify.
I'm using the `useSession()` hook in `/src/app/page.tsx`, and I've defined a custom client-side component named `Providers` which wraps the NextAuth SessionProvider, as described [here](https://nextjs.org/docs/getting-started/react-essentials#rendering-third-party-context-providers-in-server-componentsl).

**`/src/app/components/Providers.tsx`**
```
"use client";

import { SessionProvider } from "next-auth/react";
import React from "react";
import type { Session } from "next-auth";

interface ProvidersProps {
  children: React.ReactNode;
  session: Session | null;
}

function Providers({ session, children }: ProvidersProps) {
  return <SessionProvider session={session}>{children}</SessionProvider>;
}

export default Providers;
```

**`/src/app/layout.tsx`**
```
import { getServerSession } from "next-auth";
import authOptions from "./api/auth/[...nextauth]/authOptions";
import Providers from "./components/Providers";

export const metadata = {
  title: "WIP",
  description: "WIP Spotify Application",
};

export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const session = await getServerSession(authOptions);

  return (
    <html lang="en">
      <body>
        <Providers session={session}>
          {children}
        </Providers>
      </body>
    </html>
  );
}
```

**`/src/app/page.tsx`**
```
"use client";

import { getSession, signIn, signOut, useSession } from "next-auth/react";

async function Page() {
  const { data: session, status } = useSession();

  let view;

  if (status === "authenticated") {
    view = (
      <>
        Signed in as {session.user.name} <br />
        <button onClick={() => signOut()}>Sign Out</button>
      </>
    );
  } else {
    view = (
      <>
        Not signed in <br />
        <button onClick={() => signIn("spotify")}>Sign In</button>
      </>
    );
  }

  return <>{view}</>;
}

export default Page;
```

Of course, I've also defined the authentication route in `/src/app/api/auth/[...nextauth]/`:

**`/src/app/api/auth/[...nextauth]/authOptions.ts`**
```
import SpotifyProvider from "next-auth/providers/spotify";

const authOptions = {
  providers: [
    SpotifyProvider({
      clientId: process.env.SPOTIFY_CLIENT_ID as string,
      clientSecret: process.env.SPOTIFY_CLIENT_SECRET as string,
      authorization: {
        params: { scope: "playlist-read-private playlist-modify-private" },
      },
    }),
  ],
};

export default authOptions;
```

**`/src/app/api/auth/[...nextauth]/route.ts`**
```
import NextAuth from "next-auth";
import authOptions from "./authOptions";

const handler = NextAuth(authOptions);

export { handler as GET, handler as POST };
```

When starting the dev server and loading the site, everything goes as expected:

```
- ready started server on 0.0.0.0:3000, url: http://localhost:3000
- info Loaded env from { PATH TO .env.local }
- event compiled client and server successfully in 1281 ms (311 modules)
- wait compiling...
- wait compiling /page (client and server)...
- event compiled client and server successfully in 1441 ms (798 modules)
[next-auth][warn][EXPERIMENTAL_API]
`getServerSession` is used in a React Server Component.
https://next-auth.js.org/configuration/nextjs#getServerSession}
https://next-auth.js.org/warnings#EXPERIMENTAL_API
```

![sign_in](https://github.com/nextauthjs/next-auth/assets/60618151/13d28062-0e44-4198-bc0d-3e2ed5e4aa14)

When I click the _Sign In_ button, I'm redirected to the Spotify authorization page as expected, and redirected back to `http://localhost:3000/api/auth/callback/spotify` after giving my authorization.
However, rather than seeing the Sign Out page as expected, I'm greeted with a blank page, and see the following error in the console:
```
- error node_modules/next-auth/react/index.js (119:0) @ useSession
- error Error: [next-auth]: `useSession` must be wrapped in a <SessionProvider />
    at Page (./src/app/page.tsx:13:100)
null
```

But then if I restart the server and reload the page, I see the Sign Out page with my name, indicating that the session variable is accessible by `page.tsx`, at least at first.
![sign_out png](https://github.com/nextauthjs/next-auth/assets/60618151/160f228b-024d-47ca-ae96-0f95bb04466a)

The same thing happens when I click the Sign Out button. The sign out process happens as normal, but I return to a blank page with the same error in the console.

Does anyone have any idea why the SessionProvider would seemingly disappear after a redirect, or only work the first time?
I'm not sure if this is a bug in my code, in NextAuth, or in Next itself.
Any help would be appreciated, thanks.

**Comment by TommySorensen** [comment]
Maybe related to my issue? https://github.com/nextauthjs/next-auth/issues/7634

**Comment by whisdev** [comment]
This is really awesome, thx for ur affort to our dev teams :)
